scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+# 0001 - ScannerV4 APIs definition
+
+- **Status:** Accepted
+- **Created:** [2023-05-09 Tue]
+
+## Context
+
+We are transitioning Stackrox Scanner to use [ClairCore](https://github.com/quay/claircore) as its underlying scanning engine. The goal is to align vulnerability scanning results between Quay/Clair and Stackrox. The transition will be gradual, starting with container image scanning. During this period, the existing Scanner (ScannerV2) and the new ClairCore-based Scanner (ScannerV4) will be simultaneously deployed.
+
+The move to ClairCore and the adoption of ScannerV4 bring the need for new API definitions. Also, ClairCore allows for Clair deployment modes that better support the secured cluster use-case (local scanning), horizontal scalability, and high-availability scenarios. The new APIs ideally should align with ClairCore's capabilities while minimizing changes in Central.
+
+## Decision
+
+ScannerV4 APIs will exclusively use gRPC. ScannerV4 APIs are not backward compatible with ScannerV2.
+
+ScannerV4 will offer modes of operation [akin to Clair's deployment models](https://quay.github.io/clair/howto/deployment.html). Each mode will implement different gRPC services, reflecting the underlying separation between ClairCore's [libindex](https://pkg.go.dev/github.com/quay/claircore/libindex#Libindex) and [libvuln](https://pkg.go.dev/github.com/quay/claircore/libvuln#Libvuln). They will be named "indexer" and "matcher". Both modes can be enabled concurrently.
+
+ScannerV4 will use "index reports" and "vulnerability reports" data models, similar to ClairCore's. Both types will link to a "scannable resource" (e.g., a container image) using a "hash_id" managed by the clients. It uniquely identifies the resource's manifest.
```

Comment:
One thing to watch out for with gRPC is maximum response size. Claircore's native types are poorly suited to being naively serialized.

Reply:
A terminology note: what claircore calls "hash" in a lot of spots is usually spelled "digest" in the wider ecosystem (e.g. OCI Image Manifests).

Reply:
I looked into how we can increase and manage the gRPC size limits on the server side. We could use Central's gRPC library, [which sets the maximum receive to 12MB](https://github.com/stackrox/stackrox/blob/74476b76b39dfe2e9cdaeecc3e9eaf262097389f/pkg/grpc/server.go#L46), or do it on our own in Scanner.

> Claircore's native types are poorly suited to being naively serialized.

I think you are talking in terms of size, or is it something else in the Claircore's data type what might be challenging to serialize using protobufs?

Reply:
Update: [Explicitly mention we are going to use stackrox/pkg/grpc](https://github.com/stackrox/stackrox/pull/5997/files/67d35c5dc1e27202272f1c1de977e68a5817c879..3ba589c5c8bf99762c5d2f8eac332ed0050997a9)

Reply:
> I would really like to avoid depending on github.com/stackrox/rox because it is bloated and we may potentially run into circular dependencies, like we have now. I'd rather we just copy it and adjust, as necessary.

I would like to know specifically how the bloating and circular dep weights against the benefits of using a common library. Common means shared maintenance, consistency, better testing/coverage, and easier to take in enhancements and updates. A copy means constant tech debt, especially considering parts of the setup must be compatible across services, such as auth, cert, metrics, and maybe more. Also, despite the pain of being affected by changes in the lib, I see the shared responsibility over the lib between Central and Scanner will likely improve the lib, and make it more robust and reliable over time.

Circular dep can be painful, but I think the only actual dep Central has on Scanner are protos, no?

Maybe you could point out specifically what are your concerns around the bloating? And also where the circular dep can be a pain?